rust/arrow/src/datatypes.rs

```diff
@@ -156,6 +157,107 @@ make_type!(UInt64Type, u64, DataType::UInt64, 64, 0u64);
 make_type!(Float32Type, f32, DataType::Float32, 32, 0.0f32);
 make_type!(Float64Type, f64, DataType::Float64, 64, 0.0f64);
 
+macro_rules! make_temporal_type {
+    ($name:ident, $native_ty:ty, $data_ty:expr, $bit_width:expr, $default_val:expr) => {
+        pub struct $name {}
+
+        impl ArrowPrimitiveType for $name {
+            type Native = $native_ty;
+
+            fn get_data_type() -> DataType {
+                $data_ty
+            }
+
+            fn get_bit_width() -> usize {
+                $bit_width
+            }
+
+            fn default_value() -> Self::Native {
+                $default_val
+            }
+        }
+    };
+}
+
+make_temporal_type!(
+    TimestampSecondType,
+    i64,
+    DataType::Timestamp(TimeUnit::Second),
+    64,
+    0i64
+);
+make_temporal_type!(
+    TimestampMillisecondType,
+    i64,
+    DataType::Timestamp(TimeUnit::Millisecond),
+    64,
+    0i64
+);
+make_temporal_type!(
+    TimestampMicrosecondType,
+    i64,
+    DataType::Timestamp(TimeUnit::Microsecond),
+    64,
+    0i64
+);
+make_temporal_type!(
+    TimestampNanosecondType,
+    i64,
+    DataType::Timestamp(TimeUnit::Nanosecond),
+    64,
+    0i64
+);
+make_temporal_type!(Date32Type, i32, DataType::Date32(DateUnit::Day), 32, 0i32);
+make_temporal_type!(
+    Date64Type,
+    i64,
+    DataType::Date64(DateUnit::Millisecond),
+    64,
+    0i64
+);
+make_temporal_type!(
+    Time32SecondType,
+    i32,
+    DataType::Time32(TimeUnit::Second),
+    32,
+    0i32
+);
+make_temporal_type!(
+    Time32MillisecondType,
+    i32,
+    DataType::Time32(TimeUnit::Millisecond),
+    32,
+    0i32
+);
+make_temporal_type!(
+    Time64MicrosecondType,
+    i64,
+    DataType::Time64(TimeUnit::Microsecond),
+    64,
+    0i64
+);
+make_temporal_type!(
+    Time64NanosecondType,
+    i64,
+    DataType::Time64(TimeUnit::Nanosecond),
+    64,
+    0i64
+);
+make_temporal_type!(
+    IntervalYearMonthType,
```

Comment:
Interval isn't yet well defined in the format, so I was unsure of whether interval types should be backed by `i32` or `i64`.